grakn/service/Session/TransactionService.py

```diff
@@ -117,17 +113,61 @@ def run_concept_method(self, concept_id, grpc_concept_method_req):
         response = self._communicator.send(tx_request)
         return response.conceptMethod_res.response
 
+    def run_concept_iter_method(self, concept_id, grpc_concept_iter_method_req):
+        return imap(lambda res: res.conceptMethod_iter_res.response, Iterator(self._communicator, RequestBuilder.start_iterating_concept_method(concept_id, grpc_concept_iter_method_req)))
+
     def explanation(self, explainable):
         """ Retrieve the explanation of a Concept Map from the server """
         tx_request = RequestBuilder.explanation(explainable)
         response = self._communicator.send(tx_request)
         return ResponseReader.ResponseReader.create_explanation(self, response.explanation_res)
 
-    def iterate(self, iterator_id):
-        request = RequestBuilder.next_iter(iterator_id)
-        response = self._communicator.send(request)
-        return response.iterate_res
-    iterate.__annotations__ = {'iterator_id': int}
+
+class Iterator(six.Iterator):
+    def __init__(self, communicator, iter_req):
+        self._id = 0
+        self._communicator = communicator
+        self._iter_req = iter_req
+        self._buffer = []
+        self._start_iterating()
```

Comment:
Is it intended that we start iterating right away? We aren't lazy anymore then right?

Reply:
If we don't immediately send the query then define and insert queries, which are not expected to have results, will not get executed (since currently the iterators are not executed).

We can choose to use a batch size of 0 for the first iteration, but that would introduce an extra round trip to gets where we are actually expecting a result.

Also, in general, being lazy is undesirable on the network, since latency is nearly always part of the critical path of the operation (or at least, should be after optimisation).

Reply:
ah yeah i forgot about define and insert queries, this is important.